### PR TITLE
Feat/msdk 510 checkpoint all balances

### PR DESCRIPTION
**Pull request by shuffledex**

### CODE REVIEW

File: src/api/entities/Checkpoint.ts [hunk 1]
```diff
@@ -1,6 +1,10 @@
+import { Vec } from '@polkadot/types/codec';
+import { Balance } from '@polkadot/types/interfaces';
 import BigNumber from 'bignumber.js';
+import { IdentityId, Ticker } from 'polymesh-types/types';
 
 import { Context, Entity, Identity } from '~/internal';
+import { CheckpointId } from '~/polkadot';
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:7]: Should be imported from `polymesh-types/types`

File: src/api/entities/Checkpoint.ts [hunk 2]
```diff
@@ -91,23 +96,81 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
+    } = this;
+
+    const rawTicker = stringToTicker(ticker, context);
+
+    const { entries, lastKey: next } = await requestPaginated(query.asset.balanceOf, {
+      arg: rawTicker,
+      paginationOpts,
+    });
+
+    const identityBalancesOf: { did: string; balance: BigNumber }[] = [];
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:115]: currentDidBalances

File: src/api/entities/Checkpoint.ts [hunk 3]
```diff
@@ -91,23 +96,81 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
-    const { entries, lastKey: next } = await requestPaginated(
-      context.polymeshApi.query.checkpoint.balance,
-      {
-        arg: tuple(stringToTicker(ticker, context), numberToU64(id, context)),
-        paginationOpts,
+    const balanceMulti: {
+      identityId: string;
+      param: [(Ticker | CheckpointId)[], IdentityId];
+    }[] = [];
+    const balanceOf: { identity: Identity; balance: BigNumber }[] = [];
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:137]: currentIdentityBalances

Also the type here can be `IdentityBalance[]`

File: src/api/entities/Checkpoint.ts [hunk 4]
```diff
@@ -91,23 +96,81 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
+      balanceUpdatesMultiParams
+    );
 
-    const { entries, lastKey: next } = await requestPaginated(
-      context.polymeshApi.query.checkpoint.balance,
-      {
-        arg: tuple(stringToTicker(ticker, context), numberToU64(id, context)),
-        paginationOpts,
+    const balanceMulti: {
+      identityId: string;
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:134]: did

File: src/api/entities/Checkpoint.ts [hunk 5]
```diff
@@ -91,23 +96,81 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
+      param: [(Ticker | CheckpointId)[], IdentityId];
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:135]: params

File: src/api/entities/Checkpoint.ts [hunk 6]
```diff
@@ -91,23 +96,81 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
+    const rawBalanceUpdates = await query.checkpoint.balanceUpdates.multi<Vec<CheckpointId>>(
+      balanceUpdatesMultiParams
+    );
 
-    const { entries, lastKey: next } = await requestPaginated(
-      context.polymeshApi.query.checkpoint.balance,
-      {
-        arg: tuple(stringToTicker(ticker, context), numberToU64(id, context)),
-        paginationOpts,
+    const balanceMulti: {
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:133]: checkpointBalanceMultiParams

File: src/api/entities/Checkpoint.ts [hunk 7]
```diff
@@ -91,23 +96,81 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
+      identityId: string;
+      param: [(Ticker | CheckpointId)[], IdentityId];
+    }[] = [];
+    const balanceOf: { identity: Identity; balance: BigNumber }[] = [];
+
+    rawBalanceUpdates.forEach((rawCheckpointIds, index) => {
+      const firstUpdatedCheckpoint = rawCheckpointIds.find(checkpointId =>
+        u64ToBigNumber(checkpointId).gte(id)
+      );
+      const { did: identityId, balance } = identityBalancesOf[index];
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:143]: Unnecessary rename

File: src/api/entities/Checkpoint.ts [hunk 8]
```diff
@@ -91,23 +96,81 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
 
     return {
-      data,
+      data: [
+        ...balanceMulti.map(({ identityId: did }, index) => {
+          return {
+            identity: new Identity({ did }, context),
+            balance: balanceToBigNumber(rawBalanceMulti[index]),
+          };
+        }),
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:171]:
```suggestion
        ...balanceMulti.map(({ identityId: did }, index) => ({
            identity: new Identity({ did }, context),
            balance: balanceToBigNumber(rawBalanceMulti[index]),
          })),
```

File: src/api/entities/Checkpoint.ts [hunk 9]
```diff
@@ -91,23 +96,81 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
+        });
       }
-    );
+    });
 
-    const data = entries.map(([{ args: [, identityId] }, balance]) => ({
-      identity: new Identity({ did: identityIdToString(identityId) }, context),
-      balance: balanceToBigNumber(balance),
-    }));
+    const rawBalanceMulti = await query.checkpoint.balance.multi<Balance>(
```
Comment (monitz87) [inline src/api/entities/Checkpoint.ts:160]: checkpointBalances

**Review by monitz87 — CHANGES_REQUESTED** [review]

File: src/api/entities/Checkpoint.ts [hunk 10]
```diff
@@ -118,15 +181,36 @@ export class Checkpoint extends Entity<UniqueIdentifiers> {
...
-    );
+    const identity = new Identity({ did }, context);
+
+    const rawTicker = stringToTicker(ticker, context);
+    const rawU64 = numberToU64(id, context);
+    const rawIdentityId = stringToIdentityId(did, context);
+
+    const [rawBalance, sizeBalance] = await Promise.all([
+      checkpoint.balance([rawTicker, rawU64], rawIdentityId),
+      checkpoint.balance.size([rawTicker, rawU64], rawIdentityId),
```
[the remark this answers is not in the document]
Comment (VictorVicente) [reply src/api/entities/Checkpoint.ts:205]: I thought the same thing but it doesn't work. `isEmpty` returns true if balance is 0, no matters if the value exists or not. 
I have tested it and I checked this behavior.
`size` is what PolkadotJS recommends to solve these cases.

[the remark this answers is not in the document]
Comment (monitz87) [reply src/api/entities/Checkpoint.ts:205]: My bad. `isEmpty` should work when 0 isn't a reasonable value

**Review by monitz87 — APPROVED** [review]

**Pull request merged**